DCN/src/cpu/dcn_v2_cpu.cpp

```diff
@@ -1,5 +1,6 @@
 #include <vector>
```

Comment:
Hello MatthewHowe !pytorch1.8.0+cuda11.1 had launched at 2021.3.6 ,and i find that they moved THCBlas away,the THCudaBlas_Sgemm and THCudaBlas_SgemmBatched can't be use anymore.I found THCudaBlas_Sgemm can be replaced by at::cuda::blas::gemm<scalar_t>,but I don't find out how to replace THCudaBlas_SgemmBatched ....(https://github.com/pytorch/pytorch/pull/49725)

Reply:
You have to replace all `at::cuda::blas` with `torch.matmul`. Mostly same as in the CPU version.
You can check my fork https://github.com/tteepe/DCNv2, it works with PyTorch 1.8 GPU & CPU.

Reply:
@tteepe I'm testing your fork and the tests are failing. Zero offset failed and also there is runtime error for `check_gradient_dconv()`. Do you know any fixes?

Reply:
Are you running on CPU or GPU? Could you share the errors?

Reply:
@tteepe I'm running tests for GPU:

```
❯ python ../tests/test_cuda.py >> errors.txt
/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py:379: UserWarning: Input #0 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex.
  warnings.warn(
/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py:379: UserWarning: Input #2 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex.
  warnings.warn(
/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py:379: UserWarning: Input #1 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex.
  warnings.warn(
/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py:379: UserWarning: Input #3 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex.
  warnings.warn(
/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py:379: UserWarning: Input #4 requires gradient and is not a double precision floating point or complex. This check will likely fail if all the inputs are not of double precision floating point or complex.
  warnings.warn(
Traceback (most recent call last):
  File "../tests/test_cuda.py", line 310, in <module>
    check_gradient_dconv()
  File "../tests/test_cuda.py", line 102, in check_gradient_dconv
    gradcheck(
  File "/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 468, in gradcheck
    checkIfNumericalAnalyticAreClose(a, n, j)
  File "/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 449, in checkIfNumericalAnalyticAreClose
    return fail_test(error_str + 'Jacobian mismatch for output %d with respect to input %d,\n'
  File "/home/ubuntu/.pyenv/versions/3.8.6/lib/python3.8/site-packages/torch/autograd/gradcheck.py", line 367, in fail_test
    raise RuntimeError(msg)
RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[ 0.0482,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.1630,  0.4700,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.1120,  0.1633,  0.2644,  ...,  0.0000,  0.0000,  0.0000],
        ...,
        [ 0.0000,  0.0000,  0.0000,  ..., -0.0073,  0.0899, -0.0317],
        [ 0.0000,  0.0000,  0.0000,  ..., -0.3312,  0.0000, -0.0292],
        [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000]],
       device='cuda:0')
analytical:tensor([[ 0.0484,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.1630,  0.4696,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.1107,  0.1641,  0.2640,  ...,  0.0000,  0.0000,  0.0000],
        ...,
        [ 0.0000,  0.0000,  0.0000,  ..., -0.0073,  0.0896, -0.0317],
        [ 0.0000,  0.0000,  0.0000,  ..., -0.3318,  0.0000, -0.0293],
        [ 0.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000]],
       device='cuda:0')
```

The print outs for the tests are:

```
❯ cat errors.txt
torch.Size([2, 64, 128, 128])
torch.Size([20, 32, 7, 7])
torch.Size([20, 32, 7, 7])
torch.Size([20, 32, 7, 7])
0.971507, 1.943014
0.971507, 1.943014
Zero offset failed
tensor([[[[ 8.3014e-01, -5.9995e-01,  3.3635e-01,  9.7973e-01],
          [-3.9212e-01, -1.4908e+00, -7.4743e-01, -1.2212e+00],
          [-2.6852e-01,  6.1768e-01, -4.6663e-02, -8.0118e-01],
          [ 5.4885e-01, -6.4019e-01, -3.2594e-01,  5.0614e-01]],

         [[ 4.3672e-01, -1.2259e+00, -7.5590e-01,  1.1371e+00],
          [ 1.5166e-03, -4.0051e-01, -6.4815e-01,  7.1022e-01],
          [-1.1470e+00, -2.6621e+00, -9.3397e-02, -3.2880e-01],
          [-7.3536e-01, -5.0394e-01, -2.7695e-01, -7.5032e-01]]],


        [[[ 1.1741e+00, -2.4331e-01,  2.1694e-02,  1.1321e+00],
          [ 4.2356e-01,  9.0400e-01,  4.8025e-01,  8.6040e-01],
          [-9.0097e-02, -1.1146e+00,  2.7222e-01, -8.4943e-01],
          [ 1.8496e+00, -7.6862e-01,  7.1716e-01, -8.0398e-01]],

         [[-1.1821e+00, -1.2773e+00, -1.1843e-01, -1.3768e-01],
          [ 1.6133e+00, -1.2187e+00,  1.1415e-01, -1.1589e+00],
          [-8.9244e-02, -3.6727e-01, -1.8207e+00, -3.6395e-01],
          [ 1.7848e+00,  7.1422e-01,  1.4804e+00,  1.5632e+00]]]],
       device='cuda:0')
tensor([[[[ 8.3008e-01, -6.0010e-01,  3.3643e-01,  9.7949e-01],
          [-3.9209e-01, -1.4912e+00, -7.4756e-01, -1.2217e+00],
          [-2.6855e-01,  6.1768e-01, -4.6661e-02, -8.0127e-01],
          [ 5.4883e-01, -6.4014e-01, -3.2593e-01,  5.0635e-01]],

         [[ 4.3677e-01, -1.2256e+00, -7.5586e-01,  1.1367e+00],
          [ 1.5163e-03, -4.0039e-01, -6.4795e-01,  7.1045e-01],
          [-1.1475e+00, -2.6621e+00, -9.3384e-02, -3.2886e-01],
          [-7.3535e-01, -5.0391e-01, -2.7686e-01, -7.5049e-01]]],


        [[[ 1.1738e+00, -2.4329e-01,  2.1698e-02,  1.1318e+00],
          [ 4.2358e-01,  9.0381e-01,  4.8022e-01,  8.6035e-01],
          [-9.0088e-02, -1.1143e+00,  2.7222e-01, -8.4961e-01],
          [ 1.8496e+00, -7.6855e-01,  7.1729e-01, -8.0420e-01]],

         [[-1.1816e+00, -1.2773e+00, -1.1841e-01, -1.3770e-01],
          [ 1.6133e+00, -1.2188e+00,  1.1414e-01, -1.1592e+00],
          [-8.9233e-02, -3.6719e-01, -1.8203e+00, -3.6401e-01],
          [ 1.7852e+00,  7.1436e-01,  1.4805e+00,  1.5635e+00]]]],
       device='cuda:0', grad_fn=<MulBackward0>)
check_gradient_dpooling: True
```

Reply:
Okay, these are just small numerical errors. I added more generous tolerances in my latest commit. Test still may be a bit flaky due to the random input.

Reply:
@tteepe thanks for the feedback. I wasn't really sure since the tests did pass back when I used pytorch 1.7 with older gpus. Your fork has helped me so much in using CenterNet for newer gpus so thanks! ☺️

Reply:
Hi @tteepe, I still had these issues:

```
../src/cuda/dcn_v2_cuda.cu(162): error: identifier "THCudaBlas_SgemmBatched" is undefined

../src/cuda/dcn_v2_cuda.cu(309): error: identifier "THCudaBlas_Sgemm" is undefined

../src/cuda/dcn_v2_cuda.cu(359): error: argument of type "long *" is incompatible with parameter of type "int64_t *"
```
My system: Anaconda with Python 3.6.12, Windows 10, Cuda 11.1.1, Pytorch 1.8

Reply:
@rathaROG are you using his fork? those functions aren't used anymore

Reply:
@haruishi43 Thanks for your reply! Can you help verify this? This is what I did:
```
git clone https://github.com/CharlesShang/DCNv2.git
cd DCNv2
git remote add tteepe https://github.com/tteepe/DCNv2.git
git fetch tteepe
git checkout origin/master
python setup.py build develop
```
I also made some changes in `dcn_v2_im2col_cuda.cu` and `dcn_v2_psroi_pooling_cuda.cu`:
`ceil()` to `ceilf()`
`floor()` to `floorf()`
`round()` to `roundf()`

My system: windows 10, cuda 11.1.1, cudnn 8.1.1.33, anaconda python 3.6.12 with these packages:
```
Cython @ file:///C:/ci/cython_1614014892888/work
cython-bbox==0.1.3
torch==1.8.0
torchaudio==0.8.0
torchvision==0.9.0
```

All errors:
```
cpu\dcn_v2_cpu.cpp(82): error C3861: 'THFloatBlas_gemm': identifier not found
cpu\dcn_v2_cpu.cpp(101): error C3861: 'THFloatBlas_gemm': identifier not found
cpu\dcn_v2_cpu.cpp(176): error C3861: 'THFloatBlas_gemm': identifier not found
cpu\dcn_v2_cpu.cpp(216): error C3861: 'THFloatBlas_gemm': identifier not found
cpu\dcn_v2_cpu.cpp(224): error C3861: 'THFloatBlas_gemv': identifier not found
cuda/dcn_v2_cuda.cu(107): error: identifier "THCState_getCurrentStream" is undefined
cuda/dcn_v2_cuda.cu(126): error: identifier "THCudaBlas_SgemmBatched" is undefined
cuda/dcn_v2_cuda.cu(273): error: identifier "THCudaBlas_Sgemm" is undefined
cuda/dcn_v2_cuda.cu(279): error: identifier "THCState_getCurrentStream" is undefined
cuda/dcn_v2_cuda.cu(324): error: identifier "THCudaBlas_Sgemv" is undefined
```

What did I miss? 
Please help me. I really want to make it work.

Reply:
@rathaROG

tteepe's fork is actually a JIT compiled version so you don't have to do `python setup.py`.
What I to test that his fork is working is:
```
git clone  https://github.com/tteepe/DCNv2.git
cd DCN
python ../tests/test_cuda.py
```

Reply:
Hi @haruishi43, I still had one more problem:
```
[4/4] "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64/link.exe" dcn_v2_cuda.o dcn_v2_cpu.o dcn_v2_im2col_cpu.o dcn_v2_psroi_pooling_cpu.o dcn_v2_cuda.cuda.o dcn_v2_im2col_cuda.cuda.o dcn_v2_psroi_pooling_cuda.cuda.o /nologo /DLL c10.lib c10_cuda.lib torch_cpu.lib torch_cuda_cu.lib -INCLUDE:?searchsorted_cuda@native@at@@YA?AVTensor@2@AEBV32@0_N1@Z torch_cuda_cpp.lib -INCLUDE:?warp_size@cuda@at@@YAHXZ torch.lib /LIBPATH:C:\dev\exc\Anaconda3\envs\DEFT\lib\site-packages\torch\lib torch_python.lib /LIBPATH:C:\dev\exc\Anaconda3\envs\DEFT\libs "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1\lib/x64" cudart.lib /out:DCNv2_gpu.pyd
FAILED: DCNv2_gpu.pyd
"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64/link.exe" dcn_v2_cuda.o dcn_v2_cpu.o dcn_v2_im2col_cpu.o dcn_v2_psroi_pooling_cpu.o dcn_v2_cuda.cuda.o dcn_v2_im2col_cuda.cuda.o dcn_v2_psroi_pooling_cuda.cuda.o /nologo /DLL c10.lib c10_cuda.lib torch_cpu.lib torch_cuda_cu.lib -INCLUDE:?searchsorted_cuda@native@at@@YA?AVTensor@2@AEBV32@0_N1@Z torch_cuda_cpp.lib -INCLUDE:?warp_size@cuda@at@@YAHXZ torch.lib /LIBPATH:C:\dev\exc\Anaconda3\envs\DEFT\lib\site-packages\torch\lib torch_python.lib /LIBPATH:C:\dev\exc\Anaconda3\envs\DEFT\libs "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.1\lib/x64" cudart.lib /out:DCNv2_gpu.pyd
   Creating library DCNv2_gpu.lib and object DCNv2_gpu.exp
MSVCRT.lib(loadcfg.obj) : error LNK2001: unresolved external symbol __enclave_config
DCNv2_gpu.pyd : fatal error LNK1120: 1 unresolved externals
ninja: build stopped: subcommand failed.
```

Reply:
> @rathaROG
> 
> tteepe's fork is actually a JIT-compiled version so you don't have to do `python setup.py`.
> What I did to test that his fork is working is:
> 
> ```
> git clone  https://github.com/tteepe/DCNv2.git
> cd DCN
> python ../tests/test_cuda.py
> ```

Thanks for your help so far. In case you're interested, windows-ready is here
https://github.com/rathaROG/DCNv2_Windows

Reply:
> 
@tteepe Hello, I successfully compiled your version, but when I run testcuda.py, It raise an errer
No module named 'dcnv2_gpu'
Have you ever experienced this?

Reply:
@rathaROG Have you run it successfully under Linux? My environment is ：
Python 3.6.13
torch                   1.8.0+cu111
No module named 'dCNV2_GPU' in my test  
 I would appreciate it if you have a solution
![image](https://user-images.githubusercontent.com/93109645/154101341-7876946d-87e3-4347-800e-2962427bb707.png)